### PR TITLE
Rework the optimization engine (Add O4)

### PR DESCRIPTION
**Pull request by FabioLuporini**
This introduces a new optimization engine that aims to automatically apply an "optimal sequence of transformation steps" to arbitrary multilinear forms.

I'm writing a document (paper) to clarify what I mean by "optimal sequence of transformation steps", so I won't waste my (and your) time talking about it here.

I wanna clarify a few practical aspects:

* the new optimisation engine (O4) is essentially a superset of the optimisation mode O2 (O4 "uses" O2). So despite the huge amount of additions brought by this PR, things won't change (i.e. kernels won't be transformed differently) if O2 is actually in place (e.g. current Firedrake). This is on purpose, since O4 may need more time (and more people using it) to become stable. If O4 is switched on, however, all tests (even firedrake ones) at the moment pass, so one could start "beta testing" it.
* many (most!) changes are actually a refactoring or generalisation of ancient code

### CODE REVIEW

**Comment by FabioLuporini** [comment]
@miklos1 I have rebased this, I think, after almost a month, it's time to merge it

**Comment by FabioLuporini** [comment]
Feel free to play with it, check the tests etc. It works on my local branches. I'll wait for a ping from you

File: coffee/ast_analyzer.py [hunk 1]
```diff
@@ -120,8 +120,12 @@ def place(self, sym):
         return self[sym].place if self.get(sym) else None
 
     @property
+    def all_stmts(self):
+        return set((stmt_info.stmt for stmt_info in self.values() if stmt_info.stmt))
```
Comment (miklos1) [inline coffee/ast_analyzer.py]: Do we really need double parentheses here?

File: coffee/base.py [hunk 2]
```diff
@@ -675,6 +684,15 @@ def size(self):
         return self.sym.rank or (0,)
 
     @property
```
Comment (miklos1) [inline coffee/base.py]: Why is it a property? You could get the same thing by simply reading and writing `self.core`.

File: coffee/expression.py [hunk 3]
```diff
@@ -141,19 +143,29 @@ def is_scalar(self):
     def is_tensor(self):
         return not self.is_scalar
 
+    @property
```
Comment (miklos1) [inline coffee/expression.py]: ditto

File: coffee/expression.py [hunk 4]
```diff
@@ -39,7 +39,7 @@ class MetaExpr(object):
 
     """Metadata container for a compute-intensive expression."""
 
-    def __init__(self, type, parent, loops_info, domain):
+    def __init__(self, type, parent, loops_info, domain, mode=0):
```
Comment (miklos1) [inline coffee/expression.py]: What is mode 0? What are the possible values for mode?

File: coffee/ast_analyzer.py [hunk 5]
```diff
@@ -120,8 +120,12 @@ def place(self, sym):
...
+
+    @property
     def all_places(self):
-        return set((stmt_info.place for stmt_info in self.values()))
+        return set((stmt_info.place for stmt_info in self.values() if stmt_info.place))
```
Comment (miklos1) [inline coffee/ast_analyzer.py]: ditto

File: coffee/expression.py (hunk 4 above)
Comment (miklos1) [inline coffee/expression.py]: Is it the optimisation level?

File: coffee/loop_scheduler.py [hunk 6]
```diff
@@ -215,131 +215,275 @@ def merge(self, root):
...
+
+            self._success = False
+            right = dcopy(node)
+            self._cut(right.children[1], right, 'remainder')
+
+            return left, right
+
+    class CutterSum(Cutter):
+
+        def _cut(self, node, parent, side, topsum=None):
```
Comment (miklos1) [inline coffee/loop_scheduler.py]: Why is it not a visitor?

File: coffee/loop_scheduler.py [hunk 7]
```diff
@@ -215,131 +215,275 @@ def merge(self, root):
...
+                return (split, splittable)
+            return ((node, expr_info), ())
+
+    class CutterMatch(Cutter):
+
+        def __init__(self, expr_fissioner):
+            ExpressionFissioner.Cutter.__init__(self, expr_fissioner)
+            self.matched = []
+
+        def _cut(self, node, parent, side, topsum=None):
```
Comment (miklos1) [inline coffee/loop_scheduler.py]: Why is it not a visitor?

File: coffee/loop_scheduler.py [hunk 8]
```diff
@@ -215,131 +215,275 @@ def merge(self, root):
...
+            For example, consider the expression a*b + c*d; if the expression is cut
+            into chunks containing only one operand (i.e., self.cut=1), then we have
+            precisely two chunks, /split/ = a*b, /remainder/ = c*d
+
+            If the input expression is a*b + c*d + e*f, and still self.cut=1, then we
+            have two chunks, /split/ = a*b, /remainder/ = c*d + e*f; that is,
+            /remainder/ always contains the subexpression after the fission point
+            """
+            self._success = False
+            left = dcopy(node)
```
Comment (miklos1) [inline coffee/loop_scheduler.py]: You should make `_cut` a visitor which reconstructs the tree, and then you can avoid these deep copies.

File: coffee/ast_analyzer.py (hunk 1 above)
Comment (FabioLuporini) [inline coffee/ast_analyzer.py]: no

File: coffee/expression.py (hunk 4 above)
Comment (FabioLuporini) [inline coffee/expression.py]: it's in the doc string of the constructor

Comment (miklos1) [inline coffee/expression.py]: Docstring says "suggested rewrite mode". Can you elaborate? Isn't it the optimisation level?

File: coffee/loop_scheduler.py (hunk 8 above)
Comment (FabioLuporini) [inline coffee/loop_scheduler.py]: Yes, you're right. Unfortunately there is now no time for this now

File: coffee/expression.py [hunk 9]
```diff
@@ -39,7 +39,7 @@ class MetaExpr(object):
...
         """Initialize the MetaExpr.
```
Comment (FabioLuporini) [inline coffee/expression.py]: Sure. So the rewrite mode is about an expression. The optimisation level includes various things, among which the rewrite mode level.

**Pull request merged**